### ISSUE TITLE
Expiration time does not consider timezone

### ISSUE THREAD

**Issue by xong** [SymfonyCasts/verify-email-bundle#116]
https://github.com/SymfonyCasts/verify-email-bundle/blob/7b9b1f59093dd260afa91eb6b3a220906a0fe0e2/src/VerifyEmailHelper.php#L51

`time()` is timezone independent.
Is this effect wanted or a bug?

It causes the expiration dates to differ from the server time if the timezone is not UTC.

**Comment by bocharsky-bw** [comment]
Hm, if `time()` is timezone independent, then:
```php
(new \DateTime('now', new DateTimeZone('America/New_York')))->getTimestamp() === (new \DateTime('now', new DateTimeZone('UTC')))->getTimestamp();
```

So it should not be a problem. Or see https://3v4l.org/o9gka#v8.1.6

Could you share a bit more context about the use case where you have a problem with it?

**Comment by xong** [comment]
Hi, I created a registration form with the maker-bundle: `bin/console make:registration-form`

When I use `VerifyEmailSignatureComponents::getExpiresAt()` in the generated class `EmailVerifier` (see https://github.com/symfony/maker-bundle/blob/main/src/Resources/skeleton/verifyEmail/EmailVerifier.tpl.php) my time does differ 1 or 2 hours from the server time (depending on current daylight saving time).<br>
Server timezone is `Europe/Berlin`.

```
class EmailVerifier
{
    private $verifyEmailHelper;
    private $mailer;
    private $entityManager;
    private $translation;

    // ...

    public function sendEmailConfirmation(string $verifyEmailRouteName, User $user, TemplatedEmail $email): void
    {
        $signatureComponents = $this->verifyEmailHelper->generateSignature(
            $verifyEmailRouteName,
            (string) $user->getId(),
            $user->getEmail(),
            ['id' => $user->getId()]
        );

        // differs from server time
        $expiryTime = \IntlDateFormatter::formatObject($signatureComponents->getExpiresAt());

        // ...
    }
    
    // ...
}
```

**Comment by weaverryan** [comment]
Timezones are hard. We DO originally generate the integer time since the epoch. Hake we, we then convert that into a DateTime object when we pass it into the SignatureComponents. That object does still represent the correct expiration time, but I’m not sure what timezone it may be in. 

I’m honestly not sure if there is a bug or not. We probably should have just passed the integer to SignatureComponents to avoid the whole issue.

**Comment by xong** [comment]
I dived a little bit deeper.
First of all, when you use a timestamp to initialize a DateTime the timezone is ignored.
IntlDateFormatter uses the timezone of the passed DateTime. Since it is Immutable there is no way to set the timezone with the setTimeZone-method.

Only solution i found, was to create a new object and set the timezone manually:
```php
$expiresAt = \DateTime::createFromInterface($signatureComponents->getExpiresAt());
$expiresAt->setTimezone(new \DateTimeZone(date_default_timezone_get()))
$expiryTime = \IntlDateFormatter::formatObject($expiresAt);
```

I am not sure, where the bug is. Maybe it is with DatetimeImmutable::setTimezone since it has absolutely no effect.

**Comment by jrushlow** [comment]
Just skimming through the comments, this may be related to the PHP 8.1 timezone bug. See https://github.com/SymfonyCasts/reset-password-bundle/issues/219. The bug has been resolved in `php/php-src`, we're just waiting for a release. Note this only applies to PHP `8.1`

**Comment by jrushlow** [comment]
I think it's safe to close this now as the timezone bug was fixed way back in PHP `8.1`. If anyone is still having problems with timezones and `verify-email-bundle`, please create a _new_ issue and we'll work out a fix...
